### DISCUSSION TITLE
swarm with web-socket

### DISCUSSION THREAD

**Discussion by RemonAhammad** [libp2p/rust-libp2p#4305] [Q&A] [answered]
I want to use Websocket in my libp2p project. when a ws client sends a message to the server. the server will send that message via libp2p.

so I have [initialized](https://github.com/blackshadowsoftwareltd/Modular-Peer-to-Peer-Networking-Framework/blob/681a1f827d17e32e0b9d4dd3d9eb966daff997a6/src/socket/web_socket.rs#L17C14-L17C31) the web socket in my project.
everything is good. peer-to-peer message sending is working. 

when I connect a ws client with the ws server the server is [listening](https://github.com/blackshadowsoftwareltd/Modular-Peer-to-Peer-Networking-Framework/blob/681a1f827d17e32e0b9d4dd3d9eb966daff997a6/src/socket/events/handler.rs#L19C20-L19C20) to everything about the client. The socket working perfectly for now.

then when a message is received from a ws client, the ws server is sending that via libp2p. and publish method is returning Ok(). but the message does not arrive where I sent it via libp2p.

Send Method : 
```
pub fn send_text_message_sync(
    swarm: &mut Swarm<MyBehaviour>,
    topic: &IdentTopic ) {
    let message_payload = TextPayload {...........};
    let payload = MessagePayload::Text(message_payload);
    let json = serde_json::to_string(&payload).unwrap();
    let sending = swarm.behaviour_mut().gossipsub.publish(topic.clone(), json);
    match sending {
        Ok(_) => println!("Sent"),
        Err(e) => println!("Publish error :: {:?}", e),
    }
}
```
then when I disconnect the ws from the client. then the message arrives at to libp2p receiver. that mean libp2p not sending messages. but [publish method](https://github.com/blackshadowsoftwareltd/Modular-Peer-to-Peer-Networking-Framework/blob/681a1f827d17e32e0b9d4dd3d9eb966daff997a6/src/peer_to_peer/sends/text/text.rs#L44C51-L44C58) returned Ok().

I used a [loop ( pin_mut)](https://github.com/blackshadowsoftwareltd/Modular-Peer-to-Peer-Networking-Framework/blob/681a1f827d17e32e0b9d4dd3d9eb966daff997a6/src/socket/events/handler.rs#L85) to listen to the WS events in [tokio::select{}](https://github.com/blackshadowsoftwareltd/Modular-Peer-to-Peer-Networking-Framework/blob/681a1f827d17e32e0b9d4dd3d9eb966daff997a6/src/peer_to_peer/events/event.rs#L33C9-L33C25). where libp2p also listening events.
```
loop {
        //TODO Stream
        tokio::select! {

            Ok((stream, addr)) = ws_data.listener.accept() => {
               handle_ws_connection(ws_data.state.clone(), stream, addr,swarm,topic,&local_peer_id).await;
            },

            line = stdin.select_next_some() => {
                let input=line.as_ref().expect("Stdin not to close") ;
               ...
            },
            event = swarm.select_next_some() => match event {
                SwarmEvent::Behaviour(MyBehaviourEvent::Mdns(mdns::Event::Discovered(list))) => {
               ...
                },
                SwarmEvent::Behaviour(MyBehaviourEvent::Gossipsub(gossipsub::Event::Message {
                ...
                }))=>{
                    println!("message : {:?}",message);
             },
                _ => {println!("!!!!!!!!!!!!!! Default !!!!!!!!!!!!!!!");}
            }
        }
    }
```
and I can't listen to the ws events in another thread because I have to transfer a reference of the swarm to publish a message.

no progress in my work for a few days. please help me. thanks.

**Comment by thomaseizinger** [comment]
Your problem might be that `handle_ws_connection` blocks the event loop. Try using `tokio::spawn` for that future.

**Reply by RemonAhammad** [reply]
but I can't transfer swarm to the tokio::spawn()
swarm has no copy-clone feature. and swarm can't be borrowable. 
```
       Ok((stream, addr)) = ws_data.listener.accept() => {
                let state = ws_data.state.clone();
                tokio::spawn(async move {
                    handle_ws_connection(state, stream, addr , swarm, topic, &local_peer_id).await;
                });
            },
```
the error is :
```
borrowed data escapes outside of function
`swarm` escapes the function body here
```

**Reply by thomaseizinger** [reply]
You'll need to use channels to communicate between the different tasks.

**Reply by RemonAhammad** [reply]
can you suggest to me which crates I should use for channels?

**Reply by thomaseizinger** [reply]
The `futures` crate provides some async-aware channels.

**Reply by RemonAhammad** [reply]
please give me a code snippet.

**Reply by thomaseizinger** [reply, marked as answer]
The file-sharing example uses a channel to send commands to an event-loop that polls the swarm: https://github.com/libp2p/rust-libp2p/tree/master/examples/file-sharing

Hope that helps :)
